libraries/Microsoft.Bot.Builder.Core.Extensions/TranscriptLoggerMiddleware.cs

```diff
@@ -0,0 +1,126 @@
+        private ITranscriptLogger logger;
+
+        private Queue<IActivity> transcript = new Queue<IActivity>();
+
+        /// <summary>
+        /// Middleware for logging incoming and outgoing activities to a transcript store 
+        /// </summary>
+        /// <param name="transcriptLogger"></param>
+        public TranscriptLoggerMiddleware(ITranscriptLogger transcriptLogger)
+        {
+            logger = transcriptLogger ?? throw new ArgumentNullException("TranscriptLoggerMiddleware requires a ITranscriptLogger implementation.  ");
+        }
+
+        /// <summary>
+        /// initialization for middleware turn
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="nextTurn"></param>
+        /// <returns></returns>
+        public async Task OnTurn(ITurnContext context, MiddlewareSet.NextDelegate nextTurn)
+        {
+            // log incoming activity at beginning of turn
+            if (context.Activity != null)
+            {
+                if (String.IsNullOrEmpty((String)context.Activity.From.Properties["role"]))
+                    context.Activity.From.Properties["role"] = "user";
+
+                LogActivity(CloneActivity(context.Activity));
+            }
+
+            // hook up onSend pipeline
+            context.OnSendActivities(async (ctx, activities, nextSend) =>
+            {
+                // run full pipeline
+                var responses = await nextSend();
+
+                foreach (var activity in activities)
+                {
+                    LogActivity(CloneActivity(activity));
+                }
+                return responses;
+            });
+
+            // hook up update activity pipeline
+            context.OnUpdateActivity(async (ctx, activity, nextUpdate) =>
+            {
+                // run full pipeline
+                var response = await nextUpdate();
+
+                // add Message Update activity
+                var updateActivity = CloneActivity(activity);
+                updateActivity.Type = ActivityTypes.MessageUpdate;
+                LogActivity(updateActivity);
+                return response;
+            });
+
+            // hook up delete activity pipeline
+            context.OnDeleteActivity(async (ctx, reference, nextDelete) =>
+            {
+                // run full pipeline
+                await nextDelete();
+
+                // add MessageDelete activity
+                var deleteActivity = new Activity() { Type = ActivityTypes.MessageDelete }.AsMessageDeleteActivity();
+                deleteActivity.Id = reference.ActivityId;
+                deleteActivity.From = reference.Bot;
+                deleteActivity.Conversation = reference.Conversation;
+                deleteActivity.Recipient = reference.User;
+                deleteActivity.ChannelId = reference.ChannelId;
+                deleteActivity.Timestamp = DateTimeOffset.UtcNow;
```

Comment:
The JS SDK has a static helper off `TurnContext` that does this... We should add the same helper to C# so you can use that as well.

Reply:
Oh C# has it too... You should be using `TurnContext.applyConversationReference()` and be sure to set the `incoming` flag to true.

---
In reply to: [181251042](https://github.com/Microsoft/botbuilder-dotnet/pull/410#discussion_r181251042) [](ancestors = 181251042)